sqlmesh/core/metric/renderer.py

```diff
@@ -28,48 +38,52 @@ def agg(
             where: An expression to filter sources by or a dictionary of source -> filter.
             having: An expression to filter the aggregation by.
             join_type: The method of joining derived metrics, by default this is a full join.
-            dialect: The sql dialect that where and having are if they are strings.
 
         Returns:
             A SQLGlot expression containing the aggregate query.
         """
+        where = where or {}
         group_by = ensure_collection(group_by)
-        sources: t.Dict[str, t.Set[exp.AggFunc]] = {name: set() for name in self.sources}
         formulas = []
-        where = where or {}
-
-        if not isinstance(where, dict):
-            where = exp.maybe_parse(where, dialect=dialect)
-            where = {name: where for name in sources}
+        # ensure that relative order of sources stays the same if passed in
+        sources: t.Dict[str, t.Tuple[t.Set[exp.AggFunc], t.Set[str]]] = {
+            name: (set(), set()) for name in self.sources
+        }
 
         for metric in self.metrics:
             formulas.append(metric.formula)
             for agg, tables in metric.aggs.items():
-                # TODO: sort these tables depending on relationships
-                for table_name in tables:
-                    sources[table_name].add(agg)
+                aggs, joins = sources.setdefault(tables[0], (set(), set()))
```

Comment:
Relevant to my previous comment: from reading the code it's not at all obvious that a table with index `0` has a special meaning of being a measure table.